### PR TITLE
Issue #3162 - Use Jetty specific Servlet API jar.

### PR DESCRIPTION
**Pull request by sbordet**
* Updated module-info.java to reference the "jetty.servlet.api" module.
* Updated POMs to reference the o.e.j.toolchain:jetty-servlet-api artifact.
* Removed references to jetty-schemas.jar.
* Updated attribute "org.eclipse.jetty.server.webapp.ContainerIncludeJarPattern"
  to match the new Jetty Servlet API jar.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

### CODE REVIEW

File: jetty-server/src/main/config/modules/server.mod [hunk 1]
```diff
@@ -13,8 +13,7 @@ logging
 threadpool
 
 [lib]
-lib/servlet-api-4.0.jar
-lib/jetty-schemas-4.0.jar
+lib/jetty-servlet-api-4.0.0-SNAPSHOT.jar
```
Comment (joakime) [inline jetty-server/src/main/config/modules/server.mod]: Change this to `lib/jetty-servlet-api-4.*.jar`

File: NOTICE.txt [hunk 2]
```diff
@@ -105,7 +105,7 @@ The following artifacts are CDDL + GPLv2 with classpath exception.
 
 https://glassfish.dev.java.net/nonav/public/CDDL+GPL.html
 
-org.eclipse.jetty.toolchain:jetty-schemas
+org.eclipse.jetty.toolchain:jetty-servlet-api
```
[the remark this answers is not in the document]
Comment (janbartel) [reply NOTICE.txt]: @joakime actually, our standard NOTICE file covers everything already, so that's the one that we should use.

**Comment by janbartel** [comment]
@sbordet what's happening on this pr?

**Comment by joakime** [comment]
@janbartel the original goal of this PR is to have a servlet-api jar that is JPMS friendly.
There's a "jetty" servlet api jar that is Servlet 4.0 with JPMS friendly manifest.
https://oss.sonatype.org/content/repositories/jetty-snapshots/org/eclipse/jetty/toolchain/jetty-servlet-api/4.0.1-SNAPSHOT/

Where do you see the Servlet 3.0 stuff?

File: jetty-osgi/test-jetty-osgi/src/test/config/etc/jetty.xml [hunk 3]
```diff
@@ -28,13 +28,13 @@
...
+             <New id="RequestLog" class="org.eclipse.jetty.server.handler.RequestLogHandler"/>
            </Item>
            <Item>
-             <New id="DefaultHandler" class="org.eclipse.jetty.server.handler.DefaultHandler"/>
+             <New id="Contexts" class="org.eclipse.jetty.server.handler.ContextHandlerCollection"/>
            </Item>
            <Item>
-             <New id="RequestLog" class="org.eclipse.jetty.server.handler.RequestLogHandler"/>
+             <New id="DefaultHandler" class="org.eclipse.jetty.server.handler.DefaultHandler"/>
            </Item>
```
Comment (janbartel) [inline jetty-osgi/test-jetty-osgi/src/test/config/etc/jetty.xml]: Looks like an incidental fix, but ok.

File: jetty-osgi/test-jetty-osgi/src/test/java/org/eclipse/jetty/osgi/test/TestJettyOSGiBootHTTP2Conscrypt.java [hunk 4]
```diff
@@ -99,7 +101,10 @@
         List<Option> res = new ArrayList<>();
         res.add(CoreOptions.systemProperty("jetty.alpn.protocols").value("h2,http/1.1"));
         res.add(CoreOptions.systemProperty("jetty.sslContext.provider").value("Conscrypt"));
-        
+
+        String conscryptVersion = System.getProperty("conscrypt-version");
+        Assume.assumeFalse("Missing 'conscrypt-version' system property", (conscryptVersion == null || conscryptVersion.equals("null")));
+
```
Comment (janbartel) [inline jetty-osgi/test-jetty-osgi/src/test/java/org/eclipse/jetty/osgi/test/TestJettyOSGiBootHTTP2Conscrypt.java]: Nothing to do with the servlet-api change?

File: jetty-osgi/test-jetty-osgi/src/test/java/org/eclipse/jetty/osgi/test/TestJettyOSGiBootWithAnnotations.java [hunk 5]
```diff
@@ -118,16 +119,16 @@ public void testIndex() throws Exception
...
             response = req.send();
             content = response.getContentAsString();
-            assertTrue(content.contains("<p><b>Result: <span class=\"pass\">PASS</span></p>"));
+            assertThat(content, containsString("<p><b>Result: <span class=\"pass\">PASS</span></p>"));
             
             response = client.GET("http://127.0.0.1:" + port + "/frag.html");
             content = response.getContentAsString();
-            assertTrue(content.contains("<h1>FRAGMENT</h1>"));
+            assertThat(content, containsString("<h1>FRAGMENT</h1>"));
         }
```
Comment (janbartel) [inline jetty-osgi/test-jetty-osgi/src/test/java/org/eclipse/jetty/osgi/test/TestJettyOSGiBootWithAnnotations.java]: Some incidental changes here too, nowt to do with servlet api?

File: pom.xml [hunk 6]
```diff
@@ -48,8 +48,7 @@
     <logback.version>1.2.3</logback.version>
     <spring-version>5.1.1.RELEASE</spring-version>
     <jetty-test-policy.version>1.2</jetty-test-policy.version>
-    <servlet.api.version>4.0.1</servlet.api.version>
-    <servlet.schema.version>4.0.3</servlet.schema.version>
+    <servlet.api.version>4.0.1-SNAPSHOT</servlet.api.version>
```
Comment (janbartel) [inline pom.xml]: SNAPSHOT? Shouldn't it be a released version?

File: apache-jstl/src/test/java/org/eclipse/jetty/jstl/JspConfig.java [hunk 7]
```diff
@@ -33,7 +33,7 @@ public static void init(WebAppContext context, URI baseUri, File scratchDir)
     {
         context.setAttribute("javax.servlet.context.tempdir", scratchDir);
         context.setAttribute("org.eclipse.jetty.server.webapp.ContainerIncludeJarPattern",
- ".*/javax.servlet-[^/]*\\.jar$|.*/servlet-api-[^/]*\\.jar$|.*javax.servlet.jsp.jstl-[^/]*\\.jar|.*taglibs-standard-impl-.*\\.jar");
+ ".*/javax.servlet-[^/]*\\.jar$|.*/[^/]*servlet-api-[^/]*\\.jar$|.*javax.servlet.jsp.jstl-[^/]*\\.jar|.*taglibs-standard-impl-.*\\.jar");
```
Comment (janbartel) [inline apache-jstl/src/test/java/org/eclipse/jetty/jstl/JspConfig.java]: what's the motivation for changing the pattern for the servlet-api? Should such a change be made to all other patterns too?

File: jetty-documentation/src/main/asciidoc/administration/startup/screen-http-webapp-deploy-listconfig.adoc [hunk 8]
```diff
@@ -66,7 +66,6 @@ Version Information on 11 entries in the classpath.
 Note: order presented here is how they would appear on the classpath.
       changes to the --module=name command line options will be reflected here.
  0:                    3.1.0 | ${jetty.home}/lib/servlet-api-3.1.jar
- 1:                 3.1.0.M0 | ${jetty.home}/lib/jetty-schemas-3.1.jar
```
Comment (janbartel) [inline jetty-documentation/src/main/asciidoc/administration/startup/screen-http-webapp-deploy-listconfig.adoc]: All of the documentation is now wrong - it refers to servlet 3, but it needs to refer to servlet 4.

File: pom.xml (hunk 6 above)
Comment (joakime) [reply pom.xml]: The jetty-servlet-api jar in testing at the moment.
Close to being a release.
But yes, it should be a release soon.

File: jetty-osgi/test-jetty-osgi/src/test/java/org/eclipse/jetty/osgi/test/TestJettyOSGiBootWithAnnotations.java (hunk 5 above)
Comment (joakime) [reply jetty-osgi/test-jetty-osgi/src/test/java/org/eclipse/jetty/osgi/test/TestJettyOSGiBootWithAnnotations.java]: reverted.

File: jetty-osgi/test-jetty-osgi/src/test/java/org/eclipse/jetty/osgi/test/TestJettyOSGiBootHTTP2Conscrypt.java (hunk 4 above)
Comment (joakime) [reply jetty-osgi/test-jetty-osgi/src/test/java/org/eclipse/jetty/osgi/test/TestJettyOSGiBootHTTP2Conscrypt.java]: reverted.

File: jetty-osgi/jetty-osgi-boot/src/main/resources/META-INF/services/org.eclipse.jetty.webapp.Configuration [hunk 9]
```diff
@@ -1,2 +1,3 @@
 org.eclipse.jetty.osgi.annotations.AnnotationConfiguration
 org.eclipse.jetty.osgi.boot.OSGiWebInfConfiguration
+org.eclipse.jetty.osgi.boot.OSGiMetaInfConfiguration
```
[the remark this answers is not in the document]
Comment (joakime) [reply jetty-osgi/jetty-osgi-boot/src/main/resources/META-INF/services/org.eclipse.jetty.webapp.Configuration]: reverted.

File: jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/ServiceWebAppProvider.java [hunk 10]
```diff
@@ -47,7 +47,7 @@
  */
 public class ServiceWebAppProvider extends AbstractWebAppProvider implements ServiceProvider
 {   
-    private static final Logger LOG = Log.getLogger(AbstractWebAppProvider.class);
+    private static final Logger LOG = Log.getLogger(ServiceWebAppProvider.class);
```
[the remark this answers is not in the document]
Comment (joakime) [reply jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/ServiceWebAppProvider.java]: reverted.

File: jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/ServiceContextProvider.java [hunk 11]
```diff
@@ -49,7 +49,7 @@
  */
 public class ServiceContextProvider extends AbstractContextProvider implements ServiceProvider
 { 
-    private static final Logger LOG = Log.getLogger(AbstractContextProvider.class);
+    private static final Logger LOG = Log.getLogger(ServiceContextProvider.class);
```
[the remark this answers is not in the document]
Comment (joakime) [reply jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/ServiceContextProvider.java]: reverted.

File: jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/BundleWebAppProvider.java [hunk 12]
```diff
@@ -43,7 +43,7 @@
  */
 public class BundleWebAppProvider extends AbstractWebAppProvider implements BundleProvider
 {     
-    private static final Logger LOG = Log.getLogger(AbstractWebAppProvider.class);
+    private static final Logger LOG = Log.getLogger(BundleWebAppProvider.class);
```
[the remark this answers is not in the document]
Comment (joakime) [reply jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/BundleWebAppProvider.java]: reverted.

File: jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/BundleContextProvider.java [hunk 13]
```diff
@@ -44,7 +44,7 @@
  */
 public class BundleContextProvider extends AbstractContextProvider implements BundleProvider
 {    
-    private static final Logger LOG = Log.getLogger(AbstractContextProvider.class);
+    private static final Logger LOG = Log.getLogger(BundleContextProvider.class);
```
[the remark this answers is not in the document]
Comment (joakime) [reply jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/BundleContextProvider.java]: reverted.

File: jetty-documentation/src/main/asciidoc/administration/startup/screen-http-webapp-deploy-listconfig.adoc (hunk 8 above)
Comment (joakime) [reply jetty-documentation/src/main/asciidoc/administration/startup/screen-http-webapp-deploy-listconfig.adoc]: Yup, but that documentation is wrong in `jetty-10.0.x` as well.
That can be a different commit / PR.

File: apache-jstl/src/test/java/org/eclipse/jetty/jstl/JspConfig.java (hunk 7 above)
Comment (joakime) [reply apache-jstl/src/test/java/org/eclipse/jetty/jstl/JspConfig.java]: reverted.

File: jetty-documentation/src/main/asciidoc/development/maven/jetty-maven-plugin.adoc [hunk 14]
```diff
@@ -324,7 +324,7 @@ Defaults to "true".
 Controls whether any overlaid wars are added before or after the original base resource(s) of the webapp.
 See the section on link:#using-overlaid-wars[overlaid wars] for more information.
 containerIncludeJarPattern;;
-Defaults to `.*/javax.servlet-[^/]*\.jar$|.*/servlet-api-[^/]*\.jar$|.*javax.servlet.jsp.jstl-[^/]*\.jar|.*taglibs-standard-impl-.*\.jar`.
+Defaults to `.*/javax.servlet-[^/]*\.jar$|.*/[^/]*servlet-api-[^/]*\.jar$|.*javax.servlet.jsp.jstl-[^/]*\.jar|.*taglibs-standard-impl-.*\.jar`.
```
Comment (janbartel) [inline jetty-documentation/src/main/asciidoc/development/maven/jetty-maven-plugin.adoc]: I still haven't seen an explanation of why the servlet-api pattern has had an extra [^/] added? Why is this necessary, and if it is, shouldn't it be applied to all other jars as well?

Comment (joakime) [reply jetty-documentation/src/main/asciidoc/development/maven/jetty-maven-plugin.adoc]: I thought I reverted all of those.
Hmm, that's the documentation too.

Comment (joakime) [reply jetty-documentation/src/main/asciidoc/development/maven/jetty-maven-plugin.adoc]: The old pattern `.*/servlet-api-[^/]*\.jar$` would not match against `jetty-servlet-api-4.0.2.jar`

Comment (joakime) [reply jetty-documentation/src/main/asciidoc/development/maven/jetty-maven-plugin.adoc]: and if you do a mid-path search, you want to make sure that you hit on `jetty-home/lib/jetty-servlet-api-4.0.2.jar` and not something like `jetty-home/lib/servlet-api-4.0/schemas.jar`

Comment (janbartel) [reply jetty-documentation/src/main/asciidoc/development/maven/jetty-maven-plugin.adoc]: Well, the pattern should be  .*/jetty-servlet-api-[^/]*\.jar$ then.

Comment (joakime) [reply jetty-documentation/src/main/asciidoc/development/maven/jetty-maven-plugin.adoc]: this was reverted.

**Comment by sbordet** [comment]
@janbartel I have addressed your concerns with "containerIncludeJarPattern".

File: jetty-documentation/src/main/asciidoc/administration/startup/screen-http-webapp-deploy-listconfig.adoc [hunk 15]
```diff
@@ -65,7 +65,7 @@ Jetty Server Classpath:
 Version Information on 11 entries in the classpath.
 Note: order presented here is how they would appear on the classpath.
       changes to the --module=name command line options will be reflected here.
- 0:                    3.1.0 | ${jetty.home}/lib/servlet-api-3.1.jar
+ 0:                    3.1.0 | ${jetty.home}/lib/jetty-servlet-api-4.0.2.jar
```
Comment (joakime) [inline jetty-documentation/src/main/asciidoc/administration/startup/screen-http-webapp-deploy-listconfig.adoc]: whoops. missed the version identifier before the `|`

File: jetty-documentation/src/main/asciidoc/administration/startup/startup-base-vs-home.adoc [hunk 16]
```diff
@@ -158,7 +158,7 @@ Note: order presented here is how they would appear on the classpath.
  1:      1.4.1.v201005082020 | ${jetty.base}/lib/ext/javax.mail.glassfish-1.4.1.v201005082020.jar
  2:          {VERSION} | ${jetty.base}/lib/ext/test-mock-resources-{VERSION}.jar
  3:                    (dir) | ${jetty.home}/resources
- 4:                    3.1.0 | ${jetty.home}/lib/servlet-api-3.1.jar
+ 4:                    3.1.0 | ${jetty.home}/lib/jetty-servlet-api-4.0.2.jar
```
Comment (joakime) [inline jetty-documentation/src/main/asciidoc/administration/startup/startup-base-vs-home.adoc]: ditto, the version `s/3.1.0/4.0.2/`

File: jetty-documentation/src/main/asciidoc/administration/startup/startup-overview.adoc [hunk 17]
```diff
@@ -162,7 +162,7 @@ Jetty Server Classpath:
 Version Information on 7 entries in the classpath.
 Note: order presented here is how they would appear on the classpath.
       changes to the --module=name command line options will be reflected here.
- 0:                    3.1.0 | ${jetty.home}/lib/servlet-api-3.1.jar
+ 0:                    3.1.0 | ${jetty.home}/lib/jetty-servlet-api-4.0.2.jar
```
Comment (joakime) [inline jetty-documentation/src/main/asciidoc/administration/startup/startup-overview.adoc]: here too. `s/3.1.0/4.0.2/`
